backend/modules/session_manager.py

```diff
@@ -1,27 +1,282 @@
+        filenames = self._get_filenames()
+        print(f"[SessionManager] Found {len(filenames)} files: {filenames}")
+
+        sessions_with_missing_ids: list[SessionDict] = []
+        for file in filenames:
+            session_dict: SessionDict = self._read(file)
+            if not check_valid_typed_dict(session_dict, SessionDict):
+                print(
+                    f"[SessionManager] ERROR invalid session file: {file}. Ignoring",
+                    "file.",
+                )
+                continue
+
+            if "id" not in session_dict:
+                # Generate ID after loading the rest of the files.
+                sessions_with_missing_ids.append(session_dict)
+                continue
+
+            if session_dict["id"] in self._sessions.keys():
+                print(
+                    "[SessionManager] ERROR: Session ID duplicate:",
+                    f"{session_dict['id']}. Ignoring file: {file}.",
+                )
+                continue
+
+            try:
+                session_obj = Session(session_dict, self._handle_session_update)
+            except ErrorDictException:
+                print(f"[SessionManager] ERROR: Participant ID duplicate in: {file}.")
+                continue
+            self._sessions[session_dict["id"]] = session_obj
+
+        # Create sessions for files with missing session IDs
+        for session_dict in sessions_with_missing_ids:
+            session_obj = self.create_session(session_dict)
+
+    def _get_filenames(self):
+        """Get all filenames of files in `self._session_dir`."""
+        filenames: list[str] = []
+        for filename in os.listdir(self._session_dir):
+            if isfile(join(self._session_dir, filename)):
+                filenames.append(filename)
+        return filenames
+
+    def _read(self, filename: str):
+        """Read a json file.
+
+        Parameters
+        ----------
+        filename : str
+            name of the required file inside `self._session_dir`.
+        """
+        path = join(self._session_dir, filename)
+        with open(path, "r") as file:
+            data = json.load(file)
+            return data
+
+    def _write(self, session_dict: SessionDict):
+        """Write a json file.
+
+        Parameters
+        ----------
+        session_dict : custom_types.session.SessionDict
+            session data that should be written to the self._session_dir
+            directory.
+        """
+        if "id" not in session_dict:
+            print("[SessionManager] ERROR: Cannot save session without ID.")
+            return
+
+        filename = session_dict["id"] + ".json"
+        path = join(self._session_dir, filename)
+        with open(path, "w") as file:
+            json.dump(session_dict, file, indent=4)
 
-    def update_session(self, session: Session):
-        """TODO document"""
-        pass
+    def _delete_file(self, filename: str):
```

Comment:
add a pass?

Reply:
Where exactly do you mean? All functions in session_manager.py are implemented.  